### ISSUE TITLE
Moving from .zip to .tar.gz / Add permissions to .zip

### ISSUE THREAD

**Issue by gamax92** [machiav3lli/oandbackupx#3]
A gzip compressed tar file would be a much more natural format for backups, as tar files store information like file permissions, owner, group, symlinks as core information. Most Linux distributions come with gzip and tar by default, and WinRAR, 7Zip, PeaZip, etc can open .tar.gz files easily for Windows. This would get rid of the need to correct file permissions and ownership after restoring the app's data. The tar format is very simple and there are implementations such as JTar or Apache Commons Compress, both under the Apache License.

An alternative option instead of migrating to .tar.gz, is to keep using zip files, but store the unix information inside the zip file. The upper two bites of the "external file attributes" store the unix permissions, and file type (regular, directory, symlink, device, socket, etc). A 0x7875 header is added in the "extra data" to store the unix ownership. Unfortunately, the zip implementation in Java does not give you access to the internal/external file attributes and always writes 0 for both. The implementation in Apache Commons Compress does let you handle this field though. 

http://commons.apache.org/proper/commons-compress/apidocs/org/apache/commons/compress/archivers/zip/X7875_NewUnix.html
https://unix.stackexchange.com/questions/14705/the-zip-formats-external-file-attribute

**Comment by machiav3lli** [comment]
Actually the idea is there already as you may see in the README.md but it'll take some heavy changes to apply. So it'll be possibly a change for the summer, after getting to what I'd call a stable base, then adding some new features. If you'd like, you could work on proof of concept that we can implement later! Thanks for your engagement.

**Comment by gamax92** [comment]
I did actually take a shot at implementing this, porting the Compression class to Apache Commons Compress was fairly easy, but there lies a problem that in order to have permissions preserved would require root access, and applications don't run as root.

The archiving/unarchiving code would need to be handled as a separate command, either using the tar/busybox tar command, or adding tar/zip support to oab-utils, using the zip/unzip command is basically not an option, as they're too basic and don't handle storing permissions/symlinks/ownership. Using tar command for extraction would still unfortunately need special handling. Some kind of post command to fix ownership afterwards to the correct user/app id.

Alternatively, the archiving/unarchiving code can stay in the Compression class, but needs a pre/post command to gather/restore file stats, this is probably the easier option, oab-utils could generate a list of information for the Compression class to read for archiving, and the Compression class could generate the same kind of list for oab-utils to restore. In the absence of oab-utils/legacy mode, `busybox ls -nAR --full-time` could be used, and a shell script could be written to do the restoration

**Comment by machiav3lli** [comment]
As said, it'll need some special treatment, and while my focus is still on bringing a more stable base that'll be more of side job which is not a priority for the app now. Even though, feel free to make a merge request if the changes outside of the compression context could be kept minimal. Thanks

**Comment by machiav3lli** [comment]
With the latest release I introduced encryption as an option for the data backup; I think this should solve the problem permissions, as a more flexible option than giving them to a user which could change based on the Device/ROM. 

if you find this enough to solve the problem then please close the issue ticket, otherwise let me know your opinion. Thanks.

**Comment by gamax92** [comment]
What exactly does encryption have to do with file permissions?

**Comment by machiav3lli** [comment]
As I understood it, perspectively, the problem was that the permissions were too open so that manipulating the files was possible and though encryption as a security measure to avoid this type of manipulation. Otherwise constraining the permissions of files (too much) could result a problem moving backups from one device to another and using them, which a critical point of function for the app. But maybe am missing something, you can gladly illustrate it more.

**Comment by gamax92** [comment]
It's not a problem of what the permissions of the backup is, having encrypted backups is definitely a plus, but the problem was what the unix file permissions and unix file ownership of the original app files in /data were before and after. For example for your application, before backup:

All the files are owned by the app, except for the .../com.machiav3lli.backup/lib symlink, which is owned by root.
All the directories have a unix permission set of 771, except for the app directory itself, which has a unix permission of 700.
None of the files have a permission set of 771, .../com.machiav3lli.backup/files/oab-utils has a permission of 760, and the two xml files in .../com.machiav3lli.backup/files/shared_prefs have permissions of 660

When OAndBackupX and the original do a data folder restore, they do a blanket chown and chmod (or using the respective commands on oab-utils) over the entire app's data folder, which results in all files/directories/symlinks/whatever having a unix ownership of the app, and a unix permission set of 771. lib is no longer is owned by root and has too little unix permissions, all of the files have too many unix permissions.

The original point of this issue was to switch to .tar.gz which is a unix based archiving format, and natively stores unix permissions and ownership. Or to properly store unix permissions and ownership inside the zip files the backup creates. Both Java's zip encoder and zip4j seem to just write all zeros for the zip's attribute fields. I had mentioned Apache Commons Compress specifically cause it exposes the zip's internal and external file attributes fields as java class fields and methods, which is where the unix permissions would normally be stored. I know Java's zip encoder doesn't expose this, and I don't know about zip4j. Apache Commons Compress also has classes related to storing unix ownership in the zip's extra data field, X7875_NewUnix

**Comment by machiav3lli** [comment]
So considering everthing, there's couple of ways to do it:
* Super Hacky way: saving permissions in a file and reading from it after unzipping
* Less Hacky way: making a .tar before compressing and unarchive it after unzipping
* Changing OAB-utils logic to preserve permissions
* Moving to tar.gz

I'll see what I can do as soon as am over with the rewriting, cause starting from android 11 there'll be no legacy-mode for storage permissions, and so I'll have to rethink the whole structure(still don't have that much of time for this, at least not before summer/august). As said before, if you'd think outof an implementation on OAB-utils' or the app's side, I'd be glad to implement it. Otherwise I'll include this issue in the rewrite design.

**Comment by gamax92** [comment]
I haven't worked on this too much lately, I know that since apps themselves can't run as root, only the processes they spawn, my idea was actually more along the lines of the super hacky way, though with a small change.

Running oab-utils or running the `stat` command over all the files, would generate a file with a list of unix file ownership and permissions. Instead of just directly storing this in the archive, the backup app code could read the file and integrate the data into it's respective archive structures, whether it's the various fields in a tar file or the attribute and extra data fields of a zip.

Then on restoring, the backup app reads the archive and generates the same unix permission/ownership file and passes it along to oab-utils, or generates a shell script to fix all the permissions when oab-utils is unavailable. Some form of native binary usage is required regardless since that's what ends up running as superuser, not the Java code. And you need superuser access to get all the file contents and unix permissions anyway.

**Comment by TjlHope** [comment]
Just a warning as it's related to permissions, I recently had to manually backup and restore my termux data.
Although I used tar, termux was broken afterwards. Turns out android is doing some wacky stuff with selinux labling, so just preserving permissions probably isn't going to be enough, you also need to preserve security contexts and potentially extended attributes.
GNU tar can do this using --selinux and --xattrs but busybox tar can't AFAICT.

**Comment by TjlHope** [comment]
Just a warning as it's related to permissions, I recently had to manually backup and restore my termux data.
Although I used tar, termux was broken afterwards. Turns out android is doing some wacky stuff with selinux labling, so just preserving permissions probably isn't going to be enough, you also need to preserve security contexts and potentially extended attributes.
GNU tar can do this using --selinux and --xattrs but busybox tar can't AFAICT.
